src/Compilers/Core/Portable/SourceGeneration/IncrementalValueProvider.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Represents a provider of a value that can be transformed as part of constructing an execution pipeline
+    /// </summary>
+    /// <remarks>
+    /// This is an opaque type that cannot be used directly. Instead an <see cref="IIncrementalGenerator" />
+    /// will receive a set of value providers when constructing its execution pipeline. A set of extension methods
+    /// are then used to create transforms over the data that creates the actual pipeline.
+    /// </remarks>
+    /// <typeparam name="TValue">The type of value that this source provides access to</typeparam>
+    public readonly struct IncrementalValueProvider<TValue>
```

Comment:
Two providers:

IncrementalValueProvider<T>        -> Produces exactly one value
IncrementalValue**s**Provider<T> => Produces 0 or more values

Reply:
Consider changing the doc comment on the types to use the words "single" and "multiple". That makes it much clearer that the intent is about the count of the items. Really easy to read past the pluralization of the words.